### PR TITLE
Rpcn 0.5.0

### PR DESCRIPTION
**Pull request by RipleyTom**
-Contains the changes from Velocity's PR: https://github.com/RPCS3/rpcs3/pull/10547 rebased on Rpcn 0.4.0.
-Refactored np_handler(now everything is under a "np" namespace, the class implementation has been split into several files and a few header only objects have been added).
-Huge rework of the memory allocator for the match2 events data with an allocator that manages relocations when copying to game memory. This means that SceNp out of memory errors after playing a lot shouldn't happen anymore.
-Added a cache manager and an implementation of most *Local( np functions

Test server is: test-np.rpcs3.net (doesn't require email validation/token).

### CODE REVIEW

File: rpcs3/Emu/Cell/Modules/sceNp2.h [hunk 1]
```diff
@@ -1413,7 +1413,7 @@ struct SceNpMatching2RoomMessageInfo
 	u8 padding[2];
 	vm::bptr<SceNpMatching2RoomMessageDestination> dst;
 	vm::bptr<SceNpUserInfo2> srcMember;
-	vm::bcptr<void> msg;
+	vm::bptr<void> msg;
```
Comment (Megamouse) [inline rpcs3/Emu/Cell/Modules/sceNp2.h:1416]: ? seems to be correct ?

File: rpcs3/Emu/NP/fb_helpers.cpp [hunk 2]
```diff
@@ -3,431 +3,489 @@
...
-			prev_member->next = member_info;
+			room_info->roomGroupNum = resp->roomGroup()->size();
+			auto* ptr_groups        = edata.allocate<SceNpMatching2RoomGroup>(sizeof(SceNpMatching2RoomGroup) * room_info->roomGroupNum, room_info->roomGroup);
+			RoomGroups_to_SceNpMatching2RoomGroups(resp->roomGroup(), ptr_groups);
 		}
-		else
+
+		room_info->memberList.membersNum = static_cast<u32>(resp->memberList()->size());
+
+		SceNpMatching2RoomMemberDataInternal* prev_member;
```
Comment (Megamouse) [inline rpcs3/Emu/NP/fb_helpers.cpp:203]: uninitialized

File: rpcs3/Emu/NP/np_helpers.cpp [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+	void string_to_avatar_url(const std::string& str, SceNpAvatarUrl* avatar_url)
+	{
+		strcpy_trunc(avatar_url->data, str);
+	}
+
+	void string_to_communication_id(const std::string& str, SceNpCommunicationId* comm_id)
+	{
+		strcpy_trunc(comm_id->data, str);
+	}
+}
```
Comment (Megamouse) [inline rpcs3/Emu/NP/np_helpers.cpp:46]: newline

File: rpcs3/Emu/Cell/Modules/sceNp2.h (hunk 1 above)
Comment (RipleyTom) [reply rpcs3/Emu/Cell/Modules/sceNp2.h:1416]: The constness of that part of the struct only matters from the game perspective and it breaks my memory allocator allocate function.

**Pull request merged**